internal/backend/mqtt/auth/azure_iot_hub.go

```diff
@@ -0,0 +1,108 @@
+package auth
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/tls"
+	b64 "encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang"
+	"github.com/pkg/errors"
+)
+
+type AzureIoTHubConfig struct {
+	DeviceID      string `mapstructure:"device_id"`
+	IOTHubname    string `mapstructure:"iot_hub_name"`
+	CACert        string `mapstructure:"iot_hub_ca_file"`
+	DeviceKeyFile string `mapstructure:"device_key_file"`
+}
+
+type AzureIoTHubAuthentication struct {
+	clientID  string
+	username  string
+	password  string
+	tlsConfig *tls.Config
+	config    AzureIoTHubConfig
+}
+
+func createSASToken(uri string, saKey []byte) string {
+	encoded := url.QueryEscape(uri)
+	now := time.Now().Unix()
+	day := 60 * 60 * 24
+	ts := now + int64(day)
+	expiry := strconv.Itoa(int(ts))
+	signature := encoded + "\n" + expiry
+	b64key, err := b64.StdEncoding.DecodeString(string(saKey))
+	if err != nil {
+		log.Panicf("Azure IoT Key cannot be decoded")
+	}
+
+	mac := hmac.New(sha256.New, b64key)
+	mac.Write([]byte(signature))
+	hash := url.QueryEscape(b64.StdEncoding.EncodeToString(mac.Sum(nil)))
+
+	// IoT Hub SAS Token only needs `sr`, `sig` and `se` unlike other Azure servies
+	token := fmt.Sprintf("SharedAccessSignature sr=%s&sig=%s&se=%s",
+		encoded,
+		hash,
+		expiry,
+	)
+
+	return token
+}
+
+func NewAzureIoTHubAuthentication(config AzureIoTHubConfig) (Authentication, error) {
+	tlsConfig, err := newTLSConfig(config.CACert, "", "")
+	if err != nil {
+		return nil, errors.Wrap(err, "read ca cert")
+	}
+
+	deviceKey, err := ioutil.ReadFile(config.DeviceKeyFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading device key file")
+	}
+
+	username := fmt.Sprintf("%s.azure-devices.net/%s",
+		config.IOTHubname,
+		config.DeviceID,
+	)
+
+	resourceUri := fmt.Sprintf("%s.azure-devices.net/devices/%s",
+		config.IOTHubname,
+		config.DeviceID,
+	)
+
+	token := createSASToken(resourceUri, deviceKey)
+
+	return &AzureIoTHubAuthentication{
+		clientID:  config.DeviceID,
+		username:  username,
+		password:  token,
+		tlsConfig: tlsConfig,
+		config:    config,
+	}, nil
+}
+
+func (a *AzureIoTHubAuthentication) Init(opts *mqtt.ClientOptions) error {
+	broker := fmt.Sprintf("ssl://%s.azure-devices.net:8883", a.config.IOTHubname)
+	opts.AddBroker(broker)
+	opts.SetClientID(a.clientID)
+	opts.SetUsername(a.username)
+	opts.SetPassword(a.password)
+
+	return nil
+}
+
+func (a *AzureIoTHubAuthentication) Update(opts *mqtt.ClientOptions) error {
+	return nil
```

Comment:
Probably you want to move the setting of credentials / tokens to this method. What happens is that after the `ReconnectAfter` interval, the LoRa Gateway Bridge will re-connect, but before (re)connecting it will call `Update` to (re)set the credentials.

As the generated token is valid for 24 hours and only set once, this would cause the re-connect to fail. I think I should:

- [ ] remove the `password` field from the struct
- [ ] move `token := createSASToken(resourceUri, deviceKey)` to this method (so that it generates a new token ever (re)connect
- [ ] move `opts.SetPassword(token)` to this method to update the MQTT opts

Making the 24 hours configurable make this easier to test :-) E.g. you could set it to 30 seconds and see the LoRa Gateway Bridge re-connect each 30 seconds with a new token.

Reply:
✔️ remove the password field from the struct
✔️ move `token := createSASToken(resourceUri, deviceKey)` ...
✔️ move `opts.SetPassword(token)` ...